src/data/set/function.lean

```diff
@@ -1253,3 +1332,60 @@ update_comp_eq_of_not_mem_range' g a h
 lemma insert_inj_on (s : set α) : sᶜ.inj_on (λ a, insert a s) := λ a ha b _, (insert_inj ha).1
 
 end function
+
+/-! ### Equivalences, permutations -/
+
+namespace set
+variables {p : β → Prop} [decidable_pred p] {f : α ≃ subtype p} {g : perm α} {s t : set α}
+
+protected lemma maps_to.extend_domain (h : maps_to g s t) :
+  maps_to (g.extend_domain f) (coe ∘ f '' s) (coe ∘ f '' t) :=
+by { rintro _ ⟨a, ha, rfl⟩, exact ⟨_, h ha, by rw extend_domain_apply_image⟩ }
+
+protected lemma surj_on.extend_domain (h : surj_on g s t) :
+  surj_on (g.extend_domain f) (coe ∘ f '' s) (coe ∘ f '' t) :=
+begin
+  rintro _ ⟨a, ha, rfl⟩,
+  obtain ⟨b, hb, rfl⟩ := h ha,
+  exact ⟨_, ⟨_, hb, rfl⟩, by rw extend_domain_apply_image⟩,
+end
+
+protected lemma bij_on.extend_domain (h : set.bij_on g s t) :
+  bij_on (g.extend_domain f) (coe ∘ f '' s) (coe ∘ f '' t) :=
+⟨h.maps_to.extend_domain, (g.extend_domain f).injective.inj_on _, h.surj_on.extend_domain⟩
+
+end set
+
+namespace equiv
+variables (e : α ≃ β) {s : set α} {t : set β}
+
+lemma bij_on' (h₁ : maps_to e s t) (h₂ : maps_to e.symm t s) : bij_on e s t :=
+⟨h₁, e.injective.inj_on _, λ b hb, ⟨e.symm b, h₂ hb, apply_symm_apply _ _⟩⟩
+
+protected lemma bij_on (h : ∀ a, e a ∈ t ↔ a ∈ s) : bij_on e s t :=
+e.bij_on' (λ a, (h _).2) $ λ b hb, (h _).1 $ by rwa apply_symm_apply
+
+lemma inv_on (s : set α) (t : set β) : inv_on e e.symm t s :=
+⟨e.right_inverse_symm.left_inv_on _, e.left_inverse_symm.left_inv_on _⟩
+
+lemma bij_on_image : bij_on e s (e '' s) := (e.injective.inj_on _).bij_on_image
+lemma bij_on_symm_image : bij_on e.symm (e '' s) s := e.bij_on_image.symm $ e.inv_on _ _
+
+variables [decidable_eq α] {a b : α}
+
+lemma bij_on_swap (ha : a ∈ s) (hb : b ∈ s) : bij_on (swap a b) s s :=
+begin
+  refine ⟨λ x hx, _, (swap a b).injective.inj_on _, λ x hx, _⟩,
+  { obtain rfl | hxa := eq_or_ne x a,
+    { rwa swap_apply_left },
+    obtain rfl | hxb := eq_or_ne x b,
+    { rwa swap_apply_right },
+    { rwa swap_apply_of_ne_of_ne hxa hxb } },
+  { obtain rfl | hxa := eq_or_ne x a,
+    { exact ⟨_, hb, swap_apply_right _ _⟩ },
+    obtain rfl | hxb := eq_or_ne x b,
+    { exact ⟨_, ha, swap_apply_left _ _⟩ },
+    { exact ⟨_, hx, swap_apply_of_ne_of_ne hxa hxb⟩ } }
+end
+
+end equiv
```

Comment:
I think these results make more sense in `equiv/set/image`, since then you can use ` `equiv.image_eq_preimage` to golf the proof of `biJ_on_swap`.

Reply:
Whoops sorry, I forgot to apply my golf using `equiv.bij_on`.